src/middlewared/middlewared/plugins/nfs.py

```diff
@@ -77,6 +102,67 @@ class Config:
         Bool('managed_nfsd', default=True),
     )
 
+    @private
+    def setup_directories(self):
+        '''
+        We are moving the NFS state directory from /var/lib/nfs to
+        the system dataset: /var/db/system/nfs.
+        When setup_directories is called /var/db/system/nfs is expected to exist.
+
+        If STATEDIR is empty, then this might be an initialization
+        and there might be current info in /var/lib/nfs.
+
+        We always make sure the expected directories are present
+        '''
+        def name_to_id_conversion(usrgrp):
+            ''' Accept [usr,grp], return [usr,grp] as IDs '''
+            if any(isinstance(owner, str) for owner in usrgrp):
+                # Convert from name to ID, then apply
+                try:
+                    usrgrp[0] = self.middleware.call_sync('user.get_builtin_user_id', usrgrp[0])
+                    usrgrp[1] = self.middleware.call_sync('group.get_builtin_group_id', usrgrp[1])
+                except Exception as e:
+                    self.logger.error(
+                        "%s: Failed to apply ownership. usr=%r, grp=%r %r", usrgrp[0], usrgrp[1], e
+                    )
+            return usrgrp
+
+        # Initialize the system dataset NFS state directory
+        try:
+            if not any(os.scandir(NFSPath.STATEDIR.path())):
+                # System db is empty, populate it with contents of /var/lib/nfs
+                # Going forward, the system dataset should hold the NFS state data
+                if not os.path.isdir('/var/lib/nfs'):
+                    try:
+                        shutil.copytree('/var/lib/nfs', NFSPath.STATEDIR.path())
+                    except Exception as e:
+                        self.logger.error('Failed to initialize NFS state from /var/lib/nfs: %r', e)
+                    # Continue anyway.
+        except Exception as e:
+            self.logger.error("Could not find required path: %r", e)
+
+        # Make sure we have the necessary directories
+        for p in NFSPath:
+            path = p.path()
+            usrgrp = name_to_id_conversion([p.owner().get('uid'), p.owner().get('gid')])
+            try:
+                os.chmod(path, p.mode())
+                stat = os.stat(path)
+                if (stat.st_uid != usrgrp[0]) or (stat.st_gid != usrgrp[1]):
+                    self.logger.info("%s: Fix up owner for path", path)
+                    os.chown(path, usrgrp[0], usrgrp[1])
+            except FileNotFoundError:
+                if p.is_dir():
+                    os.mkdir(path, p.mode())
+                    os.chown(path, usrgrp[0], usrgrp[1])
+
+        procfs_path = '/proc/fs/nfsd/nfsv4recoverydir'
+        try:
+            with open(procfs_path, 'r+') as fp:
+                fp.write(f'{NFSPath.V4RECOVERYDIR.path()}\n')
+        except FileNotFoundError:
```

Comment:
Sorry, I most be missing something obvious so bear with me 😄  I'm still not following why you're only catching `FileNotFoundError`? Previously, you checked for files existence and then wrote to it catching any exception that may come up during the write I/O.

Now you're only catching if the file doesn't exist and then logging an error but you will crash here if writing to it fails.

What I'm getting at is, does it matter if this file exists or not?

Reply:
I don't think it matters if the file exists.   
This file was used in support of an early solution for maintaining client state.  It was supplanted with the database managed by `nfsdcld`.  However, the code supporting the old mechanism was not removed.  Since it's not yet been removed, I figure it's best to 'manage' it.
Side note:  More recently, `nfsdcld` itself has been replaced with `nfsdcltrk`, but Debian still uses `nfsdcld`.   Currently, all three are in Debian, but to my knowledge only `nfsdcld` is really 'active'.
I'll change to trap on any error.  It's more safe.